src/Persisters/Filesystem.php

```diff
@@ -73,31 +81,23 @@ public function __construct(string $path, bool $history = false, ?Serializer $se
      */
     public function save(Persistable $persistable) : void
     {
-        if (!is_file($this->path) and !is_writable(dirname($this->path))) {
-            throw new RuntimeException('Folder does not exist or is not writable');
-        }
-
-        if (is_file($this->path) and !is_writable($this->path)) {
-            throw new RuntimeException("File {$this->path} is not writable.");
-        }
-
-        if ($this->history and is_file($this->path)) {
+        if ($this->history and $this->filesystem()->has($this->path)) {
             $timestamp = (string) time();
 
             $filename = $this->path . '-' . $timestamp . '.' . self::HISTORY_EXT;
 
             $num = 0;
 
-            while (file_exists($filename)) {
+            while ($this->filesystem()->has($filename)) {
                 $filename = $this->path . '-' . $timestamp . '-' . ++$num . '.' . self::HISTORY_EXT;
             }
 
-            if (!rename($this->path, $filename)) {
+            if (!$this->filesystem()->rename($this->path, $filename)) {
                 throw new RuntimeException('Failed to create history file.');
             }
         }
 
-        $this->serializer->serialize($persistable)->write($this->path);
+        $this->serializer->serialize($persistable)->write($this->path, $this->filesystem());
```

Comment:
You can return the data from the returned Encoding object to give to Flysystem in case you decided to go with a separate Flysystem class.

```php
$data = $encoding->data();
```